modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java

```diff
@@ -0,0 +1,1691 @@
+                            log.debug("Service thread was interrupted [name=" + srvcCtx.name() + ", execId=" +
+                                srvcCtx.executionId() + ']');
+                    }
+                    catch (IgniteException e) {
+                        if (e.hasCause(InterruptedException.class) ||
+                            e.hasCause(IgniteInterruptedCheckedException.class)) {
+                            if (log.isDebugEnabled())
+                                log.debug("Service thread was interrupted [name=" + srvcCtx.name() +
+                                    ", execId=" + srvcCtx.executionId() + ']');
+                        }
+                        else {
+                            U.error(log, "Service execution stopped with error [name=" + srvcCtx.name() +
+                                ", execId=" + srvcCtx.executionId() + ']', e);
+                        }
+                    }
+                    catch (Throwable e) {
+                        U.error(log, "Service execution stopped with error [name=" + srvcCtx.name() +
+                            ", execId=" + srvcCtx.executionId() + ']', e);
+
+                        if (e instanceof Error)
+                            throw (Error)e;
+                    }
+                    finally {
+                        // Suicide.
+                        exe.shutdownNow();
+                    }
+                }
+            });
+        }
+    }
+
+    /**
+     * @param cfg Service configuration.
+     * @return Copy of service.
+     * @throws IgniteCheckedException If failed.
+     */
+    @SuppressWarnings("deprecation")
+    private Service copyAndInject(ServiceConfiguration cfg) throws IgniteCheckedException {
+        Marshaller m = ctx.config().getMarshaller();
+
+        if (cfg instanceof LazyServiceConfiguration) {
+            byte[] bytes = ((LazyServiceConfiguration)cfg).serviceBytes();
+
+            Service srvc = U.unmarshal(m, bytes, U.resolveClassLoader(null, ctx.config()));
+
+            ctx.resource().inject(srvc);
+
+            return srvc;
+        }
+        else {
+            Service srvc = cfg.getService();
+
+            try {
+                byte[] bytes = U.marshal(m, srvc);
+
+                Service cp = U.unmarshal(m, bytes, U.resolveClassLoader(srvc.getClass().getClassLoader(), ctx.config()));
+
+                ctx.resource().inject(cp);
+
+                return cp;
+            }
+            catch (IgniteCheckedException e) {
+                U.error(log, "Failed to copy service (will reuse same instance): " + srvc.getClass(), e);
+
+                return srvc;
+            }
+        }
+    }
+
+    /**
+     * @param ctxs Contexts to cancel.
+     * @param cancelCnt Number of contexts to cancel.
+     */
+    private void cancel(Iterable<ServiceContextImpl> ctxs, int cancelCnt) {
+        for (Iterator<ServiceContextImpl> it = ctxs.iterator(); it.hasNext(); ) {
+            cancel(it.next());
```

Comment:
Yes, it's expected. It's done to avoid a hang in case of node stopping or on disconnect during deployment. Now it considers behaviour as it was before, to avoid breaking existing tests. It is covered by `GridServiceProcessorStopSelfTest`.
Also, here is a task to improve this in near future:
https://issues.apache.org/jira/browse/IGNITE-9076